---
float32[] Q
float32[] R